bin/jobserver

```diff
@@ -0,0 +1,453 @@
+#!/usr/bin/env python
+# Copyright 2022 Gentoo Authors
+
+import ipaddress
+import sys
+import socket
+
+import argparse
+import threading
+import grp
+import os
+import pwd
+import datetime
+import multiprocessing
+import platform
+from socketserver import ThreadingTCPServer, BaseRequestHandler
+from time import sleep
+
+import portage.process
+from portage.util._ctypes import find_library, LoadLibrary
+
+
+__address__ = "127.0.0.1"
+__port__ = 25721
+__max_jobs__ = 1
+__job_delay__ = 0
+__min_free_mem__ = 2097152
+__max_load__ = None
+__parent_pid__ = None
+__server__ = None
+__primary_server__ = None
+
+
+__job_assign_lock__ = threading.Lock()
+__job_counter_lock__ = threading.Semaphore(10240)
+__job_free_slot_lock__ = threading.Semaphore(__max_jobs__)
+__job_running_lock__ = threading.Semaphore(__max_jobs__)
+__last_active__ = datetime.datetime.utcnow()
+__jobs_status__ = ""
+
+__parent_status__ = ""
+__show_status__ = False
+
+__has_ipv6 = None
+
+# from /usr/include/bits/sched.h
+CLONE_NEWNS = 0x00020000
+CLONE_NEWUTS = 0x04000000
+CLONE_NEWIPC = 0x08000000
+CLONE_NEWPID = 0x20000000
+CLONE_NEWNET = 0x40000000
+
+
+class JobServer(BaseRequestHandler):
+    def handle(self):
+        global __job_assign_lock__
+        global __job_counter_lock__
+        global __job_free_slot_lock__
+        global __job_running_lock__
+        global __last_active__
+
+        __job_counter_lock__.acquire()
+        __last_active__ = datetime.datetime.utcnow()
+
+        with __job_assign_lock__:
+            __job_free_slot_lock__.acquire()
+            self._allow_job()
+            __job_running_lock__.acquire()
+            self.request.sendall(b"OK")
+            sleep(__job_delay__)
+
+        while True:
+            try:
+                data = self.request.recv(10)
+            except Exception:
+                break
+            if not data:
+                break
+
+    def finish(self):
+        global __job_counter_lock__
+        global __job_free_slot_lock__
+        global __job_running_lock__
+        global __job_delay__
+
+        __job_running_lock__.release()
+        __job_counter_lock__.release()
+        __job_free_slot_lock__.release()
+
+    def _allow_job(self):
```

Comment:
Please let me note that this function is probably the heart and core of the jobserver feature. I am not saying that it is wrong how it is done currently, but I suspect that we will have a few more iterations over this function once this got merged. :)

I think the hardest part is to get the interaction with `os.getloadavg()` right, because the load average is a dampened value. This means an action, like spawning a new job, is not immediately reflected in the load average. (The same could be said to be true for `MemAvailable`.)

Therefore, I wonder if we may end up with different sleep values depending on wehter or not a job has been allowed.  Sleep longer if a new job has been granted, sleep shorter if we perform just another round of constraint checks. But of course, if that is a sensible idea has been to shown in real-world usage, and it is probably sensible to start with a simple approach.

Reply:
I did consider sleep longer to get the new job occupied resources. However, this waiting time is highly uncertain. Some tasks may take at least a few seconds to initialize, while other simple tasks may finished in less than 1 second. In this case it can lead to severe performance degradation because it cannot keep up with the running speed. 

On the other hand I do not consider available memory and system avg load as safe guards. An unreasonable amount of parallelism can totally lead to OOM. And the memory and CPU usage of a particular task is far above the average, which is out of the control of the jobserver.